internal/db/schema/migrations/postgres/14/06_wh_credential_dimension.up.sql

```diff
@@ -0,0 +1,110 @@
+begin;
+  create function _wh_distinct_credentials(p_target_id wt_public_id)
+    returns bool
+  as $$
+  declare
+    src          whx_credential_group_target%rowtype;
+    target       whx_credential_group_target%rowtype;
+  begin
+    for target in
+      select *
+        from whx_credential_group_target as t
+        where t.target_id = p_target_id
+    loop
+      select target.credential_group_key, target.credential_key, target.credential_purpose, s.* into src
+        from whx_credential_dimension_source as s
+      where s.target_id             = p_target_id
+        and s.credential_library_id = target.credential_library_id;
+
+      if src is distinct from target then
+        return true;
+      end if;
+    end loop;
+
+    return false;
+  end;
+  $$ language plpgsql;
+
+  create function wh_upsert_credential_group(p_target_id wt_public_id)
+    returns wh_dim_id
+  as $$
+  declare
+    src_count    bigint;
+    target_count bigint;
+    src           whx_credential_dimension_source%rowtype;
+    target        whx_credential_group_target%rowtype;
+    new_row       wh_credential_group%rowtype;
+    dimension_key wh_dim_id;
+  begin
+    select count(*) into target_count
+      from whx_credential_group_target as t
+      where t.target_id = p_target_id;
+
+    select count(*) into src_count
+      from whx_credential_group_source as s
+    where s.target_id = p_target_id;
+
+    if target_count != src_count or _wh_distinct_credentials(p_target_id) then
```

Comment:
If a target has 3 credentials attached to it and an attribute for only 1 of them has changed, how many new rows will be inserted into the `wh_credential_dimension` table?